src/target/llvm/llvm_instance.cc

```diff
@@ -340,9 +373,327 @@ std::string LLVMTarget::str() const {
+
+  // Check for "-" or "--".
+  if (part_this < part_end) {
+    auto& p = parts[part_this++];
+    if ((p.size() != 1 && p.size() != 2) || p.find_first_not_of('-') != std::string::npos) {
+      LOG(WARNING) << warn_header << "option must start with \"-\" or \"--\"" << warn_ignored;
+      return opt;
+    }
+  }
+
+  // Validate option name.
+  if (part_this < part_end) {
+    auto& p = parts[part_this++];
+    if (p.empty()) {
+      LOG(WARNING) << warn_header << "option name must not be empty" << warn_ignored;
+      return opt;
+    }
+    opt.name = std::move(p);
+  }
+
+  // Check type, if present.
+  Option::OptType type = Option::OptType::Invalid;
+  if (part_this < part_end) {
+    auto& p0 = parts[part_this];
+    if (p0 == ":") {
+      part_this++;  // Only advance if we saw ":".
+      if (part_this < part_end) {
+        auto& p1 = parts[part_this];
+        ICHECK(!p1.empty()) << "tokenizing error";  // This shouldn't happen.
+        if (p1 != "=") {
+          part_this++;
+          if (p1 == "bool") {
+            type = Option::OptType::Bool;
+          } else if (p1 == "int") {
+            type = Option::OptType::Int;
+          } else if (p1 == "uint") {
+            type = Option::OptType::UInt;
+          } else if (p1 == "string") {
+            type = Option::OptType::String;
+          }
+        }
+      }
+      // If there was ":", there must be a type.
+      if (type == Option::OptType::Invalid) {
+        LOG(WARNING) << warn_header << "invalid type" << warn_ignored;
+        return opt;
+      }
+    }
+  }
+
+  // Check value, if present.
+  std::optional<std::string> value;
+  if (part_this < part_end) {
+    auto& p0 = parts[part_this];
+    if (p0 == "=") {
+      part_this++;
+      if (part_this < part_end) {
+        value = std::move(parts[part_this]);
+      } else {
+        value = "";
+      }
+    } else {
+      // If there are still any parts left to be processed, there must be "=".
+      LOG(WARNING) << warn_header << "expecting \"=\"" << warn_ignored;
+      return opt;
+    }
+  }
+
+  // NOLINTNEXTLINE(runtime/int)
+  auto to_integer = [](const std::string& s) -> std::optional<long long> {
+    // std::stoll takes "long long"
+    long long number;  // NOLINT(runtime/int)
+    size_t pos;
+    try {
+      number = std::stoll(s, &pos);
+    } catch (...) {
+      return std::nullopt;
+    }
+    if (pos == s.size()) {
+      return number;
+    } else {
+      return std::nullopt;
+    }
+  };
+
+  auto to_boolean = [&to_integer](const std::string& s) -> std::optional<bool> {
+    // Return 0 or 1, if string corresponds to a valid boolean value,
+    // otherwise return 2.
+    auto ti = to_integer(s);
+    if (ti.has_value() && (ti.value() == 0 || ti.value() == 1)) {
+      return static_cast<bool>(ti.value());
+    }
+
+    std::string lower;
+    std::transform(s.begin(), s.end(), std::back_inserter(lower),
+                   [](unsigned char c) { return std::tolower(c); });
```

Comment:
could this just be std::tolower?

Reply:
`std::tolower` only does one character at a time, at least according to this page: https://en.cppreference.com/w/cpp/string/byte/tolower
Is there a different function that does a string at a time?

Reply:
If you were asking if the lambda can be replaced with just `std::tolower`, the cppreference page for it says the following about functions from \<cctype\>:
> Similarly, they should not be directly used with standard algorithms when the iterator's value type is char or signed char. Instead, convert the value to unsigned char first

Reply:
TIL :) i suppose at some point we can use std::u8string to avoid this one